### PR TITLE
build: make getting dependencies concurrently possible

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

Enable `make -j` to get dependencies for different modules concurrently where it's safe

## Commentary (optional)
51s => 33s
Screen Capture_select-area_20200604105843
Screen Capture_select-area_20200604105818

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -4,12 +4,18 @@ all:
 	$(MAKE) build
 
 .PHONY: get-deps
-get-deps:
-	pip install -r requirements.txt
+get-deps: get-deps-pip get-deps-proto get-deps-webui
 	$(MAKE) -C master $@
 	$(MAKE) -C agent $@
```
Comment (rb-determined-ai) [inline Makefile:9]: It makes sense to me that you would put these two along with `make -C proto get-deps` under a target like "get-deps-go" or something.  There's no reason, for instance, why `get-debs-webui` should be required to finish before master and agent deps.

**Comment by ybt195** [comment]
The devops-ag actually moved away from multiple targets for get-deps. In general, we want to avoid running things in parallel that would affect global caches (example being what Danny described about the go install). Even though it's probably safe, we may add things in the future that break this. We don't want people to rely on these targets for development and then remove them when we want to change the rules.

More concretely, is there a reason why we are running get-deps so often that we need to optimize it? This target should only be called when we change dependencies.

**Comment by rb-determined-ai** [comment]
I mean, it makes sense to me that different global stores (webui / python / golang) be possible to update in parallel.  If a developer doesn't want to update them in parallel that is their choice, but why wouldn't the makefile make it possible for those who do want it?

**Comment by ybt195** [comment]
@rb-determined-ai  It may be useful to bring this up in the next devops AG meeting and we can discuss it then.

File: Makefile [hunk 2]
```diff
@@ -4,12 +4,20 @@ all:
 	$(MAKE) build
 
 .PHONY: get-deps
-get-deps:
+get-deps: get-deps-pip get-deps-proto get-deps-webui get-deps-go
```
Comment (dzhu) [inline Makefile:7]: Have separate `get-deps-webui-elm` and `get-deps-webui-react` dependencies (use `subst` to turn `-` into `/`, like the other places).

Comment (hamidzr) [reply Makefile:7]: I'll look up how to use substitution. we've had issues with the node caches colliding in the past so I'd probably lean on the safer side and not separate those.

**Comment by hamidzr** [comment]
chatted with @ybt195 about the proposed change and considerations to make here. These are my main takeaways

- we are more able to fix concurrency problems with build as we have more control over it compared to dependency management systems
- different modules could now or in the future could trigger dep managers from other modules
- make -C build -j is the more important target we want people to be taking advantage of (and will get the most benefit from) and is the only officially supported target (with the current resources)

IMO the problem is partly a  `make` shortcoming. if I rely on running `make -j` on root and get-deps instability was to dissuade me from doing that and I do `make` with no `-j` I'm losing more time that I could have saved. 

considering `build` depends on `get-deps` this situation would be altogether avoided if we were to use `make get-deps && make build -j` 

Overall I think preserving the stability of `make -j` outweighs the current time saving I was seeing (maybe if I clean system caches it would be a different story :thinking: ) so I'm leaning to close this.

**Pull request closed without merging**